src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -0,0 +1,1845 @@
+                //         c.M(() => { });
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "() => { }").WithArguments("inferred delegate type").WithLocation(8, 13));
+
+            CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, expectedOutput:
+@"C.M
+C.M
+");
+        }
+
+        [Fact]
+        public void OverloadResolution_05()
+        {
+            var source =
+@"using System;
+using System.Linq.Expressions;
+class Program
+{
+    static void Report(string name, Expression e) { Console.WriteLine(""{0}: {1}"", name, e); }
+    static void F(Expression e) { Report(""F(Expression)"", e); }
+    static void F(Expression<Func<int>> e) { Report(""F(Expression<Func<int>>)"", e); }
+    static void Main()
+    {
+        F(() => 0);
+        F(() => string.Empty);
+    }
+}";
+
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular9);
+            comp.VerifyDiagnostics(
+                // (11,11): error CS8652: The feature 'inferred delegate type' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         F(() => string.Empty);
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "() => string.Empty").WithArguments("inferred delegate type").WithLocation(11, 11));
+
+            CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, expectedOutput:
+@"F(Expression<Func<int>>): () => 0
+F(Expression): () => String.Empty
+");
+        }
+
+        [Fact]
+        public void OverloadResolution_06()
+        {
+            var source =
+@"using System;
+using System.Linq.Expressions;
+class Program
+{
+    static void F(Expression e) { }
+    static void F(Expression<Func<int>> e) { }
+    static void Main()
+    {
+        F(delegate () { return 0; });
+        F(delegate () { return string.Empty; });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            // PROTOTYPE: Should report CS1946: An anonymous method expression cannot be converted to an expression tree.
+            comp.VerifyDiagnostics(
+                // (9,11): error CS1945: An expression tree may not contain an anonymous method expression
+                //         F(delegate () { return 0; });
+                Diagnostic(ErrorCode.ERR_ExpressionTreeContainsAnonymousMethod, "delegate () { return 0; }").WithLocation(9, 11),
+                // (10,11): error CS1945: An expression tree may not contain an anonymous method expression
+                //         F(delegate () { return string.Empty; });
+                Diagnostic(ErrorCode.ERR_ExpressionTreeContainsAnonymousMethod, "delegate () { return string.Empty; }").WithLocation(10, 11));
+        }
+
+        [Fact]
+        public void NullableAnalysis_01()
```

Comment:
I'm not sure what these are actually testing?

Reply:
Added comments.